shell/gpu/gpu_surface_vulkan.cc

```diff
@@ -3,9 +3,47 @@
 // found in the LICENSE file.
 
 #include "flutter/shell/gpu/gpu_surface_vulkan.h"
+#include "lib/ftl/logging.h"
 
 namespace shell {
 
-//
+GPUSurfaceVulkan::GPUSurfaceVulkan(
+    ftl::RefPtr<vulkan::VulkanProcTable> proc_table,
+    std::unique_ptr<vulkan::VulkanNativeSurface> native_surface)
+    : window_(std::move(proc_table), std::move(native_surface)),
+      weak_factory_(this) {}
+
+GPUSurfaceVulkan::~GPUSurfaceVulkan() = default;
+
+bool GPUSurfaceVulkan::Setup() {
+  // This backend does not have an explicit setup task post initialization.
+  return window_.IsValid();
+}
+
+bool GPUSurfaceVulkan::IsValid() {
+  return window_.IsValid();
+}
+
+std::unique_ptr<SurfaceFrame> GPUSurfaceVulkan::AcquireFrame(
+    const SkISize& size) {
+  auto surface = window_.AcquireSurface();
+
+  if (surface == nullptr) {
+    return nullptr;
+  }
+
+  return std::make_unique<SurfaceFrame>(
+      std::move(surface),
+      [weak_this = weak_factory_.GetWeakPtr()](SkCanvas * canvas)->bool {
+        if (canvas == nullptr || !weak_this) {
```

Comment:
By others, I am assuming you mean something like [`std::weak_ptr`](http://en.cppreference.com/w/cpp/memory/weak_ptr) with a [`lock`](http://en.cppreference.com/w/cpp/memory/weak_ptr/lock).

`ftl::WeakPtr` works a bit differently. It does not provide the same guarantees. See this comment from its header.

```
Weak pointers are not in general thread-safe. They may only be *used* on a
single thread, namely the same thread as the "originating" |WeakPtrFactory|
(which can invalidate the weak pointers that it generates).
```

IMO, the term "originating" is misleading. It is the thread on which the `WeakPtrFactory` will be collected.

In this particular case, the `WeakPtrFactory` it tied to the lifetime of the `GPUSurfaceVulkan` which is create on the platform thread. And, if successfully setup, passed-to and destroyed-by the Rasterizer (in its `Setup`) which resides on the `gl_thread`. Frame acquisition and submission also happens on the `gl_thread`. So it all works out.

As I was double checking my use though. I did notice that `ftl::WeakPtr` does not assert this behavior. I was sure it did though. I am going to try and add assertions to `ftl` as it is easy to make this mistake. I am trying to think back to all my uses of WeakPtr where I thought I was protected by this incorrect use by the assertion I thought was already there :(

FWIW, I much prefer the extra guarantee provided by obtaining the strong reference. Even at the cost of the added overhead.